### ISSUE TITLE
MaterialPage and Page Equality Broken in Navigator Comparison

### ISSUE THREAD

**Issue by rayliverified** [flutter/flutter#156551]
### Steps to reproduce

The Navigator widget's `pages` property accepts a list a pages. The issue is that the internal logic for comparing page quality doesn't compare Page equality correctly.

This results in all the pages in the navigation stack rebuilding, which is bad for performance.

**Example**

![image](https://github.com/user-attachments/assets/6e4d9fff-1bcd-482d-9375-768ee6695d36)

In this example, you can toggle between recreating the page list versus maintaining the instances of the page models. When the pages are recreated, every route is rebuilt when navigating. Watch the console for duplicate build print statements.

**Console Log**

> Add New Page 3
>
> Rebuild Page 3
> Rebuild Page 2
> Rebuild Page 1
> Rebuild Page 0

**Page Equality Tests**

Tests demonstrating Page equality is only preserved by using the same instance. Even Pages with the exact same properties are not equal.

<details><summary>Code</summary>
<p>

```dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('MaterialPage Equality Tests', () {
    test('MaterialPages with same properties should be equal', () {
      final page1 = MaterialPage(
        key: const ValueKey('test'),
        name: 'TestPage',
        arguments: {'id': 1},
        child: const Text('Test'),
      );

      final page2 = MaterialPage(
        key: const ValueKey('test'),
        name: 'TestPage',
        arguments: {'id': 1},
        child: const Text('Test'),
      );

      expect(page1 == page2, isTrue,
          reason:
              'MaterialPages with identical properties should be considered equal');
    });

    test('List of MaterialPages considers new instances as different', () {
      final List<Page> originalPages = [
        MaterialPage(
          key: const ValueKey('page1'),
          name: 'Page1',
          child: const Text('Page 1'),
        ),
        MaterialPage(
          key: const ValueKey('page2'),
          name: 'Page2',
          child: const Text('Page 2'),
        ),
      ];

      final List<Page> newPages = [
        MaterialPage(
          key: const ValueKey('page1'),
          name: 'Page1',
          child: const Text('Page 1'),
        ),
        MaterialPage(
          key: const ValueKey('page2'),
          name: 'Page2',
          child: const Text('Page 2'),
        ),
      ];

      expect(listEquals(originalPages, newPages), isTrue,
          reason:
              'Lists with new instances of MaterialPages are incorrectly considered different');
    });
  });
}
```

</p>
</details> 

### Expected results

The key expectation is that MaterialPage instances should be considered equal if they have the same key and essential properties, even if they are separate instances. This expectation aligns with how Flutter handles widget rebuilding in other contexts. For example, when a StatelessWidget or StatefulWidget is rebuilt with the same key and type, Flutter optimizes the rebuild process, understanding that the underlying structure remains the same.

In the context of navigation, this behavior is particularly important. When updating the navigation stack, developers often recreate the list of pages, perhaps adding or removing pages based on the application's state. The expectation is that if a MaterialPage in the new list has the same key and essential properties (like name, arguments, and the type of the child widget) as a page in the existing list, it should be treated as the same page. This would allow the Navigator to optimize its update process, avoiding unnecessary rebuilds of routes that haven't fundamentally changed.

The use of keys in Flutter is generally a signal for maintaining identity across rebuilds. When a developer assigns a key to a MaterialPage, they are essentially saying, "This page represents the same logical screen or route, even if I'm creating a new instance of it." The navigation system should respect this intent, using the key as the primary means of identifying and comparing pages.

I expect the Navigator to treat these two MaterialPages as the same.

```dart
          MaterialPage(
            key: ValueKey('page_1'),
            child: TestPage(title: 'Page 1'),
          ),
```

should be treated as the same page by the Navigator as 

```dart
          MaterialPage(
            key: ValueKey('page_1'),
            child: TestPage(title: 'Page 1'),
          ),
```

The keys are the same!

### Actual results

The internal logic of the Navigator does not respect the Page's Key for equality. The reason this happens is when a new MaterialPage is created, even with the same properties, it gets a new identity in memory. If it doesn't have a specific LocalKey set in the Navigator, this new identity causes it to be treated as a different page.
This means that even though two MaterialPages are equal, the Navigator might still treat them as different due to their memory identity and an existing LocalKey not being assigned.

This doesn't match the documented or expected behavior that says pages are compared via Keys.

`navigator.dart`
https://github.com/flutter/flutter/blob/82ebb74c6411afc55a8b0937535d6baeb03ea826/packages/flutter/lib/src/widgets/navigator.dart#L4173

```dart
   if (
     nextPage.key == null ||
     !pageKeyToOldEntry.containsKey(nextPage.key) ||
     !pageKeyToOldEntry[nextPage.key]!.canUpdateFrom(nextPage)
   ) {
     // Create a new entry
   } else {
     // Update existing entry
     final _RouteEntry matchingEntry = pageKeyToOldEntry.remove(nextPage.key)!;
     matchingEntry.route._updateSettings(nextPage);
     newHistory.add(matchingEntry);
   }
```

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Navigation Rebuild Test',
      home: const NavigationRebuildTest(),
    );
  }
}

class NavigationRebuildTest extends StatefulWidget {
  const NavigationRebuildTest({super.key});

  @override
  State<NavigationRebuildTest> createState() => _NavigationRebuildTestState();
}

class _NavigationRebuildTestState extends State<NavigationRebuildTest> {
  List<Page<dynamic>> currentPages = [];
  bool recreateList = true;

  @override
  void initState() {
    super.initState();
    currentPages = [
      MaterialPage(
        key: const ValueKey('page1'),
        child: const TestPage(title: 'Page 1'),
      ),
    ];
  }

  void _addPage() {
    setState(() {
      if (recreateList) {
        currentPages = [
          ...currentPages.map((page) {
            if (page is MaterialPage) {
              return MaterialPage(
                key: page.key,
                name: page.name,
                arguments: page.arguments,
                fullscreenDialog: page.fullscreenDialog,
                child: page.child,
              );
            }
            return page;
          }),
          MaterialPage(
            key: ValueKey('page${currentPages.length + 1}'),
            child: TestPage(title: 'Page ${currentPages.length + 1}'),
          ),
        ];
      } else {
        currentPages = [
          ...currentPages,
          MaterialPage(
            key: ValueKey('page${currentPages.length + 1}'),
            child: TestPage(title: 'Page ${currentPages.length + 1}'),
          ),
        ];
      }
    });
  }

  void _toggleRecreateList(bool value) {
    setState(() {
      recreateList = value;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Navigation Rebuild Test'),
        actions: [
          Row(
            children: [
              const Text('Recreate List'),
              Switch(
                value: recreateList,
                onChanged: _toggleRecreateList,
              ),
            ],
          ),
        ],
      ),
      body: Navigator(
        pages: currentPages,
        onPopPage: (route, result) {
          if (!route.didPop(result)) {
            return false;
          }
          setState(() {
            currentPages.removeLast();
          });
          return true;
        },
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _addPage,
        child: const Icon(Icons.add),
      ),
    );
  }
}

class TestPage extends StatelessWidget {
  final String title;

  const TestPage({super.key, required this.title});

  @override
  Widget build(BuildContext context) {
    debugPrint('Building $title'); // This print statement helps track rebuilds

    return Scaffold(
      appBar: AppBar(title: Text(title)),
      body: Center(child: Text(title)),
    );
  }
}
```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details open><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[√] Flutter (Channel stable, 3.24.3, on Microsoft Windows [Version 10.0.22631.4317], locale en-US)
    • Flutter version 3.24.3 on channel stable at D:\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 2663184aa7 (4 weeks ago), 2024-09-11 16:27:48 -0500
    • Engine revision 36335019a8
    • Dart version 3.5.3
    • DevTools version 2.37.3

[√] Windows Version (Installed version of Windows is version 10 or higher)

[!] Android toolchain - develop for Android devices (Android SDK version 35.0.0)
    • Android SDK at C:\Users\searc\AppData\Local\Android\sdk
    • Platform android-35, build-tools 35.0.0
    • Java binary at: C:\Program Files\Android\Android Studio\jbr\bin\java
    • Java version OpenJDK Runtime Environment (build 21.0.3+-12282718-b509.11)
    ! Some Android licenses not accepted. To resolve this, run: flutter doctor --android-licenses

[X] Chrome - develop for the web (Cannot find Chrome executable at .\Google\Chrome\Application\chrome.exe)
    ! Cannot find Chrome. Try setting CHROME_EXECUTABLE to a Chrome executable.

[√] Visual Studio - develop Windows apps (Visual Studio Community 2022 17.9.1)
    • Visual Studio at C:\Program Files\Microsoft Visual Studio\2022\Community
    • Visual Studio Community 2022 version 17.9.34616.47
    • Windows 10 SDK version 10.0.22621.0

[√] Android Studio (version 2024.2)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 21.0.3+-12282718-b509.11)

[√] VS Code (version 1.91.0)
    • VS Code at C:\Users\searc\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.98.0

[√] Connected device (3 available)
    • ONEPLUS A6013 (mobile) • 26bd994b • android-arm64  • Android 11 (API 30)
    • Windows (desktop)      • windows  • windows-x64    • Microsoft Windows [Version 10.0.22631.4317]
    • Edge (web)             • edge     • web-javascript • Microsoft Edge 129.0.2792.79

[√] Network resources
    • All expected network resources are available.
```

</details>

**Comment by rayliverified** [comment]
**TLDR:** 

`MaterialPage(key: ValueKey('page_1'))` != `MaterialPage(key: ValueKey('page_1'))`

That does not seem like the right behavior and the Navigator behavior doesn't match the documentation that says Pages with equivalent keys are treated as equal.

**Comment by nate-thegrate** [comment]
> The key expectation is that MaterialPage instances should be considered equal if they have the same key and essential properties, even if they are separate instances. This expectation aligns with how Flutter handles widget rebuilding in other contexts.

I don't believe Flutter handles widget rebuilding this way. Would you mind elaborating on this a bit?

**Comment by nate-thegrate** [comment]
Actually, maybe I should try elaborating on my point.

When a widget's key is equal to the previous key, it ensures that the widget retains the same `BuildContext` (and `State`, if applicable). But the only way to prevent rebuilds is for the current widget to be identical to the previous one.

<br>

In my opinion, MaterialPage is working as intended. Let me know if you think otherwise!

**Comment by rayliverified** [comment]
https://github.com/user-attachments/assets/e5d2f6f2-81f3-4c5e-8c62-3a957a7c5105

Thanks for taking a look and explaining your thoughts! 

I think we should give developers control over when they want pages to rebuild.
There's some "automagic" rebuild and comparator logic in the Navigator rebuild algorithm. 

In the video above, the all pages rebuild whenever a new page is added. That's very bad for performance and I don't want it to happen. However, I don't have any control over that behavior. I can't tell the Navigator to treat two equivalent pages as equivalent. Because the Navigator doesn't respect keys and the MaterialPage doesn't respect equality, I have no control over telling the pages not to rebuild.

```dart
// ignore_for_file: unnecessary_brace_in_string_interps

import 'package:flutter/material.dart';

/// Example of working Navigator comparison behavior.
/// Both default MaterialPage and Custom equality pages match.
/// When the page variable is reused, equality works.
void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Page Rebuild Test',
      home: const PageRebuildTest(),
    );
  }
}

class PageRebuildTest extends StatefulWidget {
  const PageRebuildTest({super.key});

  @override
  State<PageRebuildTest> createState() => _PageRebuildTestState();
}

class _PageRebuildTestState extends State<PageRebuildTest> {
  GlobalKey<NavigatorState> navigatorKey = GlobalKey<NavigatorState>();
  int _pageCount = 1;
  bool _optimizedList = false;

  @override
  void initState() {
    super.initState();
  }

  List<Page> buildPages(int count) {
    List<Page> pagesHolder = [];

    for (int i = 1; i <= count; i++) {
      if (_optimizedList) {
        pagesHolder.add(OptimizedMaterialPage(
          key: ValueKey('page$i'),
          child: _TestPage(key: ValueKey('page$i'), title: 'Page $i'),
        ));
      } else {
        pagesHolder.add(MaterialPage(
          key: ValueKey('page$i'),
          name: 'page$i',
          arguments: null,
          fullscreenDialog: false,
          child: _TestPage(key: ValueKey('page$i'), title: 'Page $i'),
        ));
      }
    }

    return pagesHolder;
  }

  void _addPage() {
    setState(() {
      _pageCount++;
    });
  }

  void _toggleOptimizedList(bool value) {
    setState(() {
      _optimizedList = value;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(_optimizedList ? 'Optimized Pages' : 'Material Pages'),
        actions: [
          Row(
            children: [
              const Text('Optimized List'),
              Switch(
                value: _optimizedList,
                onChanged: _toggleOptimizedList,
              ),
            ],
          ),
        ],
      ),
      body: Navigator(
        key: navigatorKey,
        pages: [
          ...buildPages(_pageCount),
        ],
        onPopPage: (route, result) {
          if (!route.didPop(result)) {
            return false;
          }
          setState(() {
            if (_pageCount > 1) _pageCount--;
          });
          return true;
        },
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _addPage,
        child: const Icon(Icons.add),
      ),
    );
  }
}

class _TestPage extends StatelessWidget {
  final String title;

  const _TestPage({super.key, required this.title});

  @override
  Widget build(BuildContext context) {
    print('Building $title'); // This print statement helps track rebuilds
    return Scaffold(
      appBar: AppBar(title: Text(title)),
      body: Center(child: Text(title)),
    );
  }
}

class OptimizedMaterialPage<T> extends Page<T> {
  final Widget child;

  const OptimizedMaterialPage({
    required this.child,
    super.key,
    super.name,
    super.arguments,
  });

  @override
  Route<T> createRoute(BuildContext context) {
    return MaterialPageRoute<T>(
      settings: this,
      builder: (BuildContext context) => child,
    );
  }

  @override
  bool operator ==(Object other) {
    if (other.runtimeType != runtimeType) return false;
    bool result = other is OptimizedMaterialPage &&
        other.key == key &&
        other.name == name &&
        other.arguments == arguments;
    debugPrint('Page Equality: $result');
    return result;
  }

  @override
  int get hashCode => Object.hash(key, name, arguments);
}
```

**Comment by nate-thegrate** [comment]
Thanks for the example & video!

Overall, I really like the solution you've proposed in the `OptimizedMaterialPage<T>` class declaration. I do have a concern though:

```dart
    if (other.runtimeType != runtimeType)
```

My understanding is that accessing an object's `runtimeType` can have a negative impact on performance; this is one of the reasons why widgets don't have equality implementations by default, and why we use functions like [`objectRuntimeType()`](https://api.flutter.dev/flutter/foundation/objectRuntimeType.html).

<br><br>

The above example does a great job of demonstrating the potential impact of equality checks on performance; at the same time, the `buildPages` method appears to have been specifically engineered to create new object instances rather than reusing them.

`List` objects are mutable by default: I bet if `pagesHolder` were a field in the `_PageRebuildTestState` class rather than a local variable, there would be an even bigger performance boost than what `OptimizedMaterialPage` currently provides.

**Comment by darshankawar** [comment]
Keeping the issue open based on the report and discussion for team's tracking.

**Comment by rayliverified** [comment]
> My understanding is that accessing an object's runtimeType can have a negative impact on performance; this is one of the reasons why widgets don't have equality implementations by default, and why we use functions like [objectRuntimeType()](https://api.flutter.dev/flutter/foundation/objectRuntimeType.html).

How to use `objectRuntimeType`?
For it to work, you still need to use a `.runtimeType` right?
```dart
  @override
  bool operator ==(Object other) {
    if (objectRuntimeType(this, 'OptimizedMaterialPage') !=
        other.runtimeType) return false;
```

> List objects are mutable by default: I bet if pagesHolder were a field in the _PageRebuildTestState class rather than a local variable, there would be an even bigger performance boost than what OptimizedMaterialPage currently provides.

The Navigator rebuilds based on list equality. If you use a cached list variable, the Navigator won't navigate. 

This is why the list is often passed as:
`List.from(_pagesHolder)`

The way I understand List equality, List quality is based on isIdentical(). Contents can change whenever and List will always == List if they are the same instance.

Next, I'm going to clearly lay out the problem, goals, and propose a path forward.

**Comment by rayliverified** [comment]
**Problem**
Navigator does not give developers control over Page equality.
This leads to bad app performance as pages rebuild when they aren't intended.
- Every page rebuilds on each navigation event.
- Pages rebuild hundreds of times on show and hide keyboard.
  - If you're on the 10th page and open the keyboard, all 10 pages will rebuild 4-6 times, for a total of 40 - 60 rebuilds. Closing the keyboard adds another 40 - 60 rebuilds. 
  - The lag is noticeable.

**Solution**
Update the Navigator to support comparing page equality with `ValueKey()`. 
This matches the expected Flutter developer experience. Aligning the Page widget rebuild behavior with other widgets.

**TODOs**
- [ ] Identify the Navigator code that does a page comparison.
- [ ] Update the Navigator code to check keys == before identity.

**Comment by nate-thegrate** [comment]
Thanks for following up on this!

<br>

> How to use `objectRuntimeType`?

The idea behind `objectRuntimeType()` is that it will use the actual `runtimeType` in debug mode but returns a pre-defined string in release mode.
This function is usually used in `toString` implementations and isn't meant for equality checks—it's just an example of a way we've put in effort to avoid referencing the `runtimeType`. Sorry for the miscommunication.

In almost any case (aside from within private or `final` classes), accessing the `runtimeType` is required when overriding the `==` operator.

<br>

> The Navigator rebuilds based on list equality. If you use a cached list variable, the Navigator won't navigate.

Fair point! My knowledge here is somewhat limited, since I've always used the Navigator's `push()` and `pop()` methods, rather than interfacing with a list of pages (and fortunately, when these methods are used, there doesn't seem to be any noticeable lag).

<br>

> This matches the expected Flutter developer experience. Aligning the Page widget rebuild behavior with other widgets.

I'm not quite sure I agree. IMO the way to align Page-based rebuilds with other parts of the framework would be to use `const` constructors when possible.

**Comment by nate-thegrate** [comment]
@rayliverified the [`Element.rebuild` documentation](https://main-api.flutter.dev/flutter/widgets/Element/rebuild.html#not-rebuilding) has some pretty cool info that perhaps you'd find interesting:

<br>

> ### Not rebuilding
> Before an [Element](https://main-api.flutter.dev/flutter/widgets/Element-class.html) is told to update itself with a new [Widget](https://main-api.flutter.dev/flutter/widgets/Widget-class.html), the old and new objects are compared using `==`.
>
> In general, this is equivalent to doing a comparison using [identical](https://main-api.flutter.dev/flutter/dart-core/identical.html) to see if the two objects are in fact the exact same instance. If they are, and if the element is not already marked dirty for other reasons, then the element skips updating itself as it can determine with certainty that there would be no value in updating itself or its descendants.
>
> It is strongly advised to avoid overriding `==` on [Widget](https://main-api.flutter.dev/flutter/widgets/Widget-class.html) objects. While doing so seems like it could improve performance, in practice, for non-leaf widgets, it results in O(N²) behavior. This is because by necessity the comparison would have to include comparing child widgets, and if those child widgets also implement `==`, it ultimately results in a complete walk of the widget tree... which is then repeated at each level of the tree. In practice, just rebuilding is cheaper. (Additionally, if any subclass of [Widget](https://main-api.flutter.dev/flutter/widgets/Widget-class.html) used in an application implements `==`, then the compiler cannot inline the comparison anywhere, because it has to treat the call as virtual just in case the instance happens to be one that has an overridden operator.)
>
> Instead, the best way to avoid unnecessary rebuilds is to cache the widgets that are returned from [State.build](https://main-api.flutter.dev/flutter/widgets/State/build.html), so that each frame the same widgets are used until such time as they change. Several mechanisms exist to encourage this: `const` widgets, for example, are a form of automatic caching (if a widget is constructed using the `const` keyword, the same instance is returned each time it is constructed with the same arguments).
>
> Another example is the [AnimatedBuilder.child](https://main-api.flutter.dev/flutter/widgets/ListenableBuilder/child.html) property, which allows the non-animating parts of a subtree to remain static even as the [AnimatedBuilder.builder](https://main-api.flutter.dev/flutter/widgets/AnimatedBuilder/builder.html) callback recreates the other components.

**Comment by rayliverified** [comment]
Thanks Nate, I was just looking for this!


This mostly explains the rebuild behavior. There's still some unexplained
behavior with the list equality add and remove behavior. There's also the
internal navigator algorithm's diff calculation which for example rebuild
the current page and the newly pushed page when navigating.

My proposed solution is to use Keys. Is there hesitation about that? How
else do we give the developer control over rebuilds?

**Comment by rayliverified** [comment]
Caveat, MaterialPage and Page aren't widgets. They're classes.

They wrap data for navigation. The Navigator calls createRoute which uses
the builder function to build the widgets.

**Comment by nate-thegrate** [comment]
> My proposed solution is to use Keys. Is there hesitation about that?

Yeah, unfortunately I feel hesitant about potential confusion if keys were able to prevent rebuilds when assigned to a Page but not anywhere else.

<br>

> How else do we give the developer control over rebuilds?

IMO the default identity-based equality would work fine. In the case of the Navigator, this would probably involve `List.from(_pagesHolder)` as you mentioned earlier.

<br>

> Caveat, MaterialPage and Page aren't widgets. They're classes.

True that 👍

**Comment by rayliverified** [comment]
> Yeah, unfortunately I feel hesitant about potential confusion if keys
were able to prevent rebuilds when assigned to a Page but not anywhere else.

Keys are commonly used by developers to prevent rebuilds and indicate to
the Flutter framework that one widget is the same as another widget.

For example, it's used when rearranging items in a column. Without
assigning a key, the tree doesn't know that the items are the same when
their order changes. However, with keys, the framework is able to switch
the render objects in the tree, instead of rebuilding.

What you're saying about keys I'm hearing for the first time. The usage of
keys I described is what's commonly understood in the broader community as
well.

What are the technical details we are missing?

**Comment by nate-thegrate** [comment]
> with keys, the framework is able to switch the render objects in the tree, instead of rebuilding.

This is half true: the framework does switch the render objects based on their keys, but the rebuilding still takes place.

**Comment by johnpryan** [comment]
Thanks for writing a detailed issue @rayliverified!

cc: @goderbauer @chunhtai do you know why this is causing the Page's child widget to rebuild? I don't have much insight into the page diffing algorithm here. It makes sense to me that using the same list of pages _should_ avoid rebuilding each page's child widget if the keys are the same. 

It looks like go_router re-builds the list of pages whenever it's state changes, so this likely affects go_router users as well.

> Pages rebuild hundreds of times on show and hide keyboard.

This sounds like a separate issue - can you provide more information about this?

**Comment by rayliverified** [comment]
Thank you @johnpryan for weighing in!

> This sounds like a separate issue - can you provide more information about this?

When you tap a TextInput and the keyboard shows, the page rebuilds. I would not expect all the other pages in the stack to rebuild as well which is what I am seeing in my app. I haven't been able to isolate it to a minimal example yet due to the complexity.
This could be triggered by a MediaQuery if you have a MediaQuery.of() that you're using on a page, say for responsiveness. It seems that pages in the stack are still "listening" to the MediaQuery when page resize events are dispatched.

This could be a separate issue or related.

**Comment by rayliverified** [comment]
I learned some more information about Keys.

![Flutter Keys Documentation](https://github.com/user-attachments/assets/5e6d215e-8808-4254-b6c6-e12bb492804f)
![Flutter Keys Documentation 2](https://github.com/user-attachments/assets/20dc68b3-1c2d-453a-ae2f-6254a6a43ba0)
![Flutter Keys Documentation 3](https://github.com/user-attachments/assets/42f14eba-1727-4d79-9fa8-ae5c1ae18663)
![Flutter Keys Documentation 4](https://github.com/user-attachments/assets/3493b503-453c-4530-a762-38450c58d171)

**Flutter Documentation**
https://docs.flutter.dev/ui#keys
https://api.flutter.dev/flutter/widgets/Widget-class.html?_gl=1*1uof9fu*_ga*MTk3MDQ5Mzc5My4xNzA4Mzg2NTkz*_ga_04YGWK0175*MTczMDA0OTgzNy4xMTYuMS4xNzMwMDQ5ODU1LjAuMC4w#key
https://api.flutter.dev/flutter/widgets/Widget/key.html
https://api.flutter.dev/flutter/foundation/Key-class.html

The current argument is that Keys should not behave the same way when used for Page widgets than other Flutter widgets. I don't think that's right. Instead, Keys behavior should be respected by the Navigator.

**Comment by nate-thegrate** [comment]
> A new widget will only be used to update an existing element if its key is the same as the key of the current widget associated with the element.

Yes, this matches my understanding. Is there any part of the documentation that talks about keys being used to prevent rebuilds?

**Comment by rayliverified** [comment]
> If the [runtimeType](https://api.flutter.dev/flutter/dart-core/Object/runtimeType.html) and [key](https://api.flutter.dev/flutter/widgets/Widget/key.html) properties of the two widgets are [operator==](https://api.flutter.dev/flutter/widgets/Widget/operator_equals.html), respectively, then the new widget replaces the old widget by updating the underlying element (i.e., by calling [Element.update](https://api.flutter.dev/flutter/widgets/Element/update.html) with the new widget). Otherwise, the old element is removed from the tree, the new widget is inflated into an element, and the new element is inserted into the tree.

What this means to me and the behavior I've seen is that when the StatefulWidget key matches, the Widget's build() function is not called, it's not rebuilt. Instead, the element continues to persist, and its state remains unchanged. The existing widget's element and state are preserved because Flutter detects that the key and runtimeType are equal, allowing it to optimize the rendering process by reusing the element. This means that the build() function isn't called again in this scenario, as there's no need to recreate the widget subtree or rebuild the interface.

**Comment by nate-thegrate** [comment]
> If the [runtimeType](https://api.flutter.dev/flutter/dart-core/Object/runtimeType.html) and [key](https://api.flutter.dev/flutter/widgets/Widget/key.html) properties of the two widgets are [operator==](https://api.flutter.dev/flutter/widgets/Widget/operator_equals.html), respectively, then the new widget replaces the old widget by updating the underlying element (i.e., by calling [Element.update](https://api.flutter.dev/flutter/widgets/Element/update.html) with the new widget).

This is true—and both [StatelessElement.update](https://api.flutter.dev/flutter/widgets/StatelessElement/update.html) and [StatefulElement.update](https://api.flutter.dev/flutter/widgets/StatefulElement/update.html) call `rebuild(force: true)`.

I don't believe there's ever a situation where the `key` prevents rebuilds: the only way to do it is by retaining the widget's identity (either with a `const` constructor or by caching the value somewhere).

Here's a DartPad example: https://dartpad.dev/?id=48bc1f0196d4026b81f12bcfeb5d9b2e

**Comment by gnprice** [comment]
This is expected behavior. It's exactly how rebuilds behave on widgets and elements, as described in the [Element.rebuild] documentation that @nate-thegrate quoted at https://github.com/flutter/flutter/issues/156551#issuecomment-2420920622 . So it's natural that pages and routes behave the same way.

> `MaterialPage` instances should be considered equal if they have the same key and essential properties, even if they are separate instances.

The same quoted documentation explains why this kind of `==` definition would be expensive on widgets and isn't recommended for them. That reasoning applies to `MaterialPage` too.

In the sample code, the existing pages are all described by new `MaterialPage` values copied from the old ones:
```dart
        currentPages = [
          ...currentPages.map((page) {
            if (page is MaterialPage) {
              return MaterialPage(
                key: page.key,
                name: page.name,
                arguments: page.arguments,
                fullscreenDialog: page.fullscreenDialog,
                child: page.child,
              );
            }
```

But in a real app, there's not much reason to copy the objects unless you're changing something. And if you're changing something, then it needs to rebuild anyway — that's the point of the change.

If you're seeing rebuilds like these because something in your app is re-computing the pages from scratch even when the answers don't change, the solution is the same strategy described for widgets in that Element.rebuild documentation: cache the Page objects somewhere, so that when they haven't changed you can return the existing ones instead.

**Comment by gnprice** [comment]
> Pages rebuild hundreds of times on show and hide keyboard.
> * If you're on the 10th page and open the keyboard, all 10 pages will rebuild 4-6 times, for a total of 40 - 60 rebuilds. Closing the keyboard adds another 40 - 60 rebuilds.

> This could be triggered by a MediaQuery if you have a MediaQuery.of() that you're using on a page, say for responsiveness. It seems that pages in the stack are still "listening" to the MediaQuery when page resize events are dispatched.

This doesn't seem to be related to comparisons or keys in the list of pages, so it's independent of the main topic of this thread. If you want to discuss it further, a fresh issue thread would be most effective.

**Comment by rayliverified** [comment]
Thank you everyone on the Flutter team for discussing this issue. I've got a lot more knowledge and clarity on the inner workings. Closing the issue at this time.

**Solution**
Cache the page instance reference and pass it back into the Navigator instead of creating a new instance each time.
See the implementation in `navigation_utils` for an example: https://github.com/rayliverified/NavigationUtils/blob/master/lib/src/navigation_builder.dart
Caching the pages improves performance by 10x in some cases, such as when there are many pages, nested routes, or layout heavy pages.
